src/effects/effectslot.cpp

```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+    for (const auto& pParameterSlot : m_parameterSlots) {
+        if (pParameterSlot->parameterType() == parameterType) {
+            if (iSlotNumber == slotNumber) {
+                return pParameterSlot;
+            }
+            ++iSlotNumber;
         }
+    }
+    return EffectParameterSlotBasePointer();
+}
+
+void EffectSlot::loadEffect(const EffectManifestPointer pManifest,
+                            std::unique_ptr<EffectProcessor> pProcessor,
+                            const QSet<ChannelHandleAndGroup>& activeChannels) {
+    unloadEffect();
 
-        emit(effectLoaded(pEffect, m_iEffectNumber));
+    m_pManifest = pManifest;
+
+    if (pManifest == EffectManifestPointer()) {
+        // No new effect to load; just unload the old effect and return.
+        emit(effectChanged());
+        return;
+    }
+
+    for (auto& parameterMapping : m_parameterSlotPositionToManifestIndex) {
+        parameterMapping.clear();
+    }
+
+    int index = 0;
+    for (const auto& pManifestParameter: m_pManifest->parameters()) {
+        EffectParameter* pParameter = new EffectParameter(
+                this, m_pEffectsManager, m_parameters.size(), pManifestParameter);
+        m_parameters.append(pParameter);
+
+        m_parameterSlotPositionToManifestIndex[pManifestParameter->parameterType()].push_back(index);
+        ++index;
+    }
+
+    addToEngine(std::move(pProcessor), activeChannels);
+
+    m_pControlLoaded->forceSet(1.0);
+
+    loadParameters();
+
+    if (m_pEffectsManager->isAdoptMetaknobValueEnabled()) {
+        slotEffectMetaParameter(m_pControlMetaParameter->get(), true);
     } else {
-        clear();
-        // Broadcasts a null effect pointer
-        emit(effectLoaded(EffectPointer(), m_iEffectNumber));
+        m_pControlMetaParameter->set(getMetaknobDefault());
+        slotEffectMetaParameter(getMetaknobDefault(), true);
     }
-    emit(updated());
+
+    emit(effectChanged());
+    updateEngineState();
 }
 
-void EffectSlot::clear() {
-    if (m_pEffect) {
-        m_pEffect->disconnect(this);
+void EffectSlot::loadParameters() {
+    int numTypes = static_cast<int> (EffectManifestParameter::ParameterType::NUM_TYPES);
+    for (int parameterTypeId=0 ; parameterTypeId<numTypes ; ++parameterTypeId) {
+        const EffectManifestParameter::ParameterType parameterType =
+                static_cast<EffectManifestParameter::ParameterType> (parameterTypeId);
+
+        m_pControlNumParameters[parameterType]->forceSet(numParameters(parameterType));
+
+        unsigned int iParameter = 0;
+        unsigned int numParameterSlots = m_iNumParameterSlots[parameterType];
+        for (int i=0 ; i<m_parameterSlotPositionToManifestIndex[parameterType].size() ; ++i) {
+            const int iParameterPosition = m_parameterSlotPositionToManifestIndex[parameterType].value(i, -1);
+
+            auto pParameter = m_parameters.value(iParameterPosition, nullptr);
+
+            VERIFY_OR_DEBUG_ASSERT(pParameter != nullptr) {
```

Comment:
I think this is because of the previously unclear naming. It should have been iParameterSlot. I have renamed it to parameterSlotIndex now.

Not incrementing the same here enables us to skip invalid parameter manifests without wasting a parameter slot.

Reply:
Okay, that makes sense. It would be good to add a comment here explaining that.